public/static/js/projects.js

```diff
@@ -1,1221 +1,1412 @@
 
-  $scope.triggerLabel = function (trigger) {
-    if (!trigger || !trigger.project) {
-      return "";
-    }
-    var out = "When a " + trigger.level;
-    out = out + " matching variant '" + (trigger.variant_regex || "*") + "'";
-    out = out + " and task '" + (trigger.task_regex || "*") + "'";
-    out = out + " in project '" + trigger.project + "'";
-    out = out + " finishes with status '" + (trigger.status || "*") + "'";
-    out = out + ", run the tasks defined in '" + trigger.file + "'";
-    if (trigger.command) {
-      out = out + " generated by '" + trigger.command + "'";
-    }
-    if (trigger.alias) {
-      out = out + " filtered to alias '" + trigger.alias + "'";
-    }
-    return out;
-  }
+    $scope.addProjectVar = function () {
+      if ($scope.proj_var.name && $scope.proj_var.value) {
+        $scope.settingsFormData.project_vars[$scope.proj_var.name] =
+          $scope.proj_var.value;
+        if ($scope.proj_var.is_private) {
+          $scope.settingsFormData.private_vars[$scope.proj_var.name] = true;
+        }
+        if ($scope.proj_var.is_restricted) {
+          $scope.settingsFormData.restricted_vars[$scope.proj_var.name] = true;
+        }
 
-  $scope.periodicBuildLabel = function (definition) {
-    if (!definition) {
-      return "";
-    }
-    if (definition.message) {
-      return definition.message;
-    }
-    return `Every ${definition.interval_hours} hours`;
-  }
+        $scope.proj_var.name = "";
+        $scope.proj_var.value = "";
+        $scope.proj_var.is_private = false;
+        $scope.proj_var.is_restricted = false;
+      }
+    };
 
-  $scope.$watch("settingsForm.$dirty", function (dirty) {
-    if (dirty) {
-      $scope.saveMessage = "You have unsaved changes.";
-      $scope.isDirty = true;
-    }
-  });
+    $scope.addGithubAlias = function () {
+      if (!$scope.validPatchDefinition($scope.github_alias)) {
+        $scope.invalidGithubPatchDefinitionMessage =
+          "A patch alias must have variant regex, and exactly one of task regex or tag";
+        return;
+      }
+      var item = Object.assign({}, $scope.github_alias);
+      item.alias = "__github";
+      $scope.settingsFormData.github_aliases = $scope.settingsFormData.github_aliases.concat(
+        [item]
+      );
+      delete $scope.github_alias;
+      $scope.invalidGitHubPatchDefinitionMessage = "";
+    };
 
-  $scope.getAlertDisplay = function (alertObj) {
-    if (alertObj.provider == 'email') {
-      return "Send an e-mail to " + alertObj.settings.recipient
-    }
-    if (alertObj.provider == 'jira') {
-      return "File a " + alertObj.settings.issue + " JIRA ticket in " + alertObj.settings.project
-    }
-    if (alertObj.provider == 'slack') {
-      return "Send a slack message to " + alertObj.settings.channel
-    }
-    return 'unknown'
-  }
+    $scope.validGithubTrigger = function (trigger_alias) {
+      found = false;
+      $scope.patch_trigger_aliases.forEach((patch_trigger_alias) => {
```

Comment:
we could probably use patch_trigger_aliases's [includes() method](https://www.w3schools.com/jsref/jsref_includes_array.asp)

Reply:
discussed over slack